Request and response compression

riz0id opened the pull request:
Draft zstandard compression feature branch.

ixmatus commented on src/Network/GRPC/MQTT/Proto.hs:
```diff
@@ -0,0 +1,672 @@
+    DRepBoolean
+  | -- | Decimal representations capture the @double@ and @float@ .proto types.
+    DRepDecimal
+  | -- | Integer representations capture all signed, unsigned, and fixed-width
```
Why is conflating these many different integer representations into one, with the `castDatumRepWith` function, safe?

I have the same question for the `DRepDecimal` and `DRepString` tags.

riz0id replied:
However, in conflating `DRepDecimal` and `DRepInteger` representations, there is nothing preventing a situation where `castDatum` can overflow or underflow in the process of casting `Int` to `Int8`. In these cases though, the casts will inherit the overflow/underflow behavior of `fromIntegral`.

ixmatus replied:
The point you make in your second comment is what I meant by asking why it was safe, you're relying on `fromIntegral`'s behavior which could produce an overflow. I don't think this is safe and instead you should explicate the different integer and decimal types in DRep so that you have a 1:1 conversion to their counterparts in Haskell (and the conversions can happen safely).

Friede80 replied:
I guess I'm just looking at it from this side of the `proto3-suite` API boundary. Here you will only ever see an `Int` inside a `DotProtoValue`.

I agree `Integer` would be a safe type to use for `IntLit`, but that seems like something that would need to be addressed within `proto3-suite`.

ixmatus replied:
I understand now @Friede80 and @riz0id and agree, without a `ProtoDatum Int8` instance, this is fine. I'm still worried about converting from a Double to a Float because that is lossy.

ixmatus commented on src/Network/GRPC/MQTT/Proto.hs:
```diff
@@ -0,0 +1,672 @@
+
+-- Proto Options - Option Errors -----------------------------------------------
+
+-- | 'ProtoOptionError' represents the errors that can improper protobuf option
```
```suggestion
-- | 'ProtoOptionError' represents the errors that improper protobuf option
```

ixmatus commented on src/Network/GRPC/MQTT/Proto.hs:
```diff
@@ -0,0 +1,672 @@
+--
+-- @since 0.1.0.0
+data ProtoOptionError
+  = -- | 'ProtoOptionAlreadySet' errors represets protobuf declarations with
```
```suggestion
  = -- | 'ProtoOptionAlreadySet' errors are emitted when protobuf declarations with
```

ixmatus commented on src/Network/GRPC/MQTT/Proto.hs:
```diff
@@ -0,0 +1,672 @@
+-- @since 0.1.0.0
+data ProtoOptionError
+  = -- | 'ProtoOptionAlreadySet' errors represets protobuf declarations with
+    -- multiple value set for the same option.
```
```suggestion
    -- multiple values are set for the same option.
```

ixmatus commented on src/Network/GRPC/MQTT/Proto.hs:
```diff
@@ -0,0 +1,672 @@
+  { serviceName :: DotProtoIdentifier
+  , serviceDefs :: [DotProtoServicePart]
+  }
+  deriving (Eq, Show)
```
```suggestion
  deriving stock (Eq, Show)
```

ixmatus commented on src/Network/GRPC/MQTT/Proto.hs:
```diff
@@ -0,0 +1,672 @@
+--
+-- @since 0.1.0.0
+methodsOf :: Applicative f => ProtoService -> (RPCMethod -> f a) -> f [a]
+methodsOf dotproto k = traverse k (getServiceMethods dotproto)
```
Just a style nit/suggestion, feel free to ignore because I think this is fine:

```suggestion
methodsOf dotproto = for (getServiceMethods dotproto)
```

ixmatus commented on src/Network/GRPC/MQTT/Proto.hs:
```diff
@@ -0,0 +1,672 @@
+--
+-- @since 0.1.0.0
+servicesOf :: Applicative f => DotProto -> (ProtoService -> f a) -> f [a]
+servicesOf dotproto k = traverse k (getFileServices dotproto)
```
Just a style nit/suggestion, feel free to ignore as I think this is fine:

```suggestion
servicesOf dotproto = for (getFileServices dotproto)
```

ixmatus commented on src/Network/GRPC/MQTT/Proto.hs:
```diff
@@ -0,0 +1,672 @@
+
+-- Proto Methods ---------------------------------------------------------------
+
+-- | Querys a 'RPCMethod' for it's gRPC method type.
```
```suggestion
-- | Querys a 'RPCMethod' for its gRPC method type.
```

ixmatus commented on src/Network/GRPC/MQTT/Proto.hs:
```diff
@@ -0,0 +1,672 @@
+-- * Manipulating the protobuf syntax trees.
+--
+-- @since 0.1.0.0
+module Network.GRPC.MQTT.Proto
```
Many of these functions could usefully be moved to `proto3-suite`. I don't suggest doing that in this PR though.

ixmatus commented on src/Network/GRPC/MQTT/RemoteClient.hs:
```diff
@@ -204,34 +203,37 @@ pattern AuxMessageTerminate = AuxControlMessage (Enumerated (Right AuxControlTer
 -- | Handles AuxControl signals from the "/control" topic
 handleControlMessage :: Logger -> SessionHandle -> LByteString -> IO ()
 handleControlMessage logger handle msg = do
```
```suggestion
handleControlMessage logger handle msg =
```

ixmatus commented on src/Network/GRPC/MQTT/RemoteClient.hs:
```diff
@@ -204,34 +203,37 @@ pattern AuxMessageTerminate = AuxControlMessage (Enumerated (Right AuxControlTer
+            _ <- tryPutTMVar (hdlHeartbeat handle) ()
+            pure ()
+        Right ctrl -> do
+          -- TODO: cancel hdlThread and abort
```
Is it the case that we need to cancel the `hdlThread` on an unknown control message? I don't think we want to cancel the thread because we may receive other control messages that we do know how to handle.

Same goes for the todo comment on the `Left err` case too.

riz0id replied:
Good point! I didn't think about that.

ixmatus commented on src/Network/GRPC/MQTT/Types.hs:
```diff
@@ -123,5 +142,5 @@ data ClientHandler where
 -- messages will not be published immediately.
-data Batched = Unbatched | Batched
-  deriving (Eq, Ord, Lift)
+-- data Batched = Unbatched | Batched
```
I would just delete this code since you've replace it with the newtype + pattern synonyms.

ixmatus commented on test/proto/service.proto:
```diff
@@ -8,29 +8,37 @@ package haskell.grpc.mqtt.test;
-  rpc BiDiStreamCall(stream BiDiRequestReply) returns (stream BiDiRequestReply) {}
+  rpc BiDiStreamCall(stream BiDiRequestReply) returns (stream BiDiRequestReply) {
+    option haskell.grpc.mqtt.client_clevel = CLEVEL_10;
+    option haskell.grpc.mqtt.server_clevel = CLEVEL_10;
```
Now that I'm seeing it in a proto, is it possible to mix and match CLevels between the client and server? Or do they need to match? If they need to match, do we need distinct server and client clevel options?

ixmatus replied:
That's pretty cool.

ixmatus commented on src/Network/GRPC/MQTT/Option/Batched.hs:
```diff
@@ -0,0 +1,175 @@
+-- | @since 0.1.0.0
+instance Enum Batched where
+  toEnum 1 = Batched
+  toEnum _ = Unbatched
```
Yeah, that makes sense to me.

Friede80 commented on src/Network/GRPC/MQTT/TH/Proto.hs:
```diff
@@ -40,39 +75,292 @@ import Proto3.Suite.DotProto.AST
+castProtoOption opt =
+  let val :: DotProtoValue
+      val = DotProto.dotProtoOptionValue opt
+   in case castDatum val of
+        Just px -> pure px
+        Nothing ->
+          let issue :: ProtoOptionError
+              issue = OptionValueTypeMismatch (datumRepOf val) opt
+           in throwError issue
```
`whenNothing`
```suggestion
   in castDatum val
         `whenNothing` throwError (OptionValueTypeMismatch (datumRepOf val) opt)
```

Friede80 commented on src/Network/GRPC/MQTT/Client.hs:
```diff
@@ -209,78 +194,112 @@ mqttRequest MQTTGRPCClient{..} baseTopic nmMethod useBatchedStream request = do
-            liftEither $ unwrapBiDiStreamResponse rsp
+    let timeout :: TimeoutSeconds
+        timeout = requestTimeout request
+     in timeoutGRPC timeout case request of
```
You can save a nesting level here by dropping the `in` since you are already inside a `do` block.

Friede80 commented on src/Network/GRPC/MQTT/Client.hs:
```diff
@@ -209,78 +194,112 @@ mqttRequest MQTTGRPCClient{..} baseTopic nmMethod useBatchedStream request = do
+
+              readResponseStream <- mkStreamRead do 
+                bytes <- withExceptT parseErrorToRCE (packetReader responseChan)
+                if Option.isServerCompressed opts
+                  then pure bytes
+                  else withExceptT Compress.toRemoteError do 
+                    fromStrict <$> Compress.decompress (toStrict bytes)
```
Are these cases backwards?

Friede80 commented on src/Network/GRPC/MQTT/Client.hs:
```diff
@@ -209,78 +194,112 @@ mqttRequest MQTTGRPCClient{..} baseTopic nmMethod useBatchedStream request = do
+
+              readResponseStream <- mkStreamRead do 
+                bytes <- withExceptT parseErrorToRCE (packetReader responseChan)
+                if Option.isServerCompressed opts
+                  then pure bytes
+                  else withExceptT Compress.toRemoteError do 
+                    fromStrict <$> Compress.decompress (toStrict bytes)
```
Are these cases backwards?

Friede80 commented on src/Network/GRPC/MQTT/Client.hs:
```diff
@@ -451,3 +466,5 @@ makeMethodRequestTopic baseTopic sid (MethodName nm) =
         Nothing -> throwIO (BadRPCMethodTopicError (decodeUtf8 nm))
         Just ts -> pure (baseTopic <> "grpc" <> "request" <> sid <> ts)
+
+-- Topics ----------------------------------------------------------------------
```
Stray section-break line?

Friede80 commented on src/Network/GRPC/MQTT/Proto.hs:
```diff
@@ -0,0 +1,672 @@
+  deriving stock (Eq, Ord)
+
+-- | @since 0.1.0.0
+instance Show ProtoIOError where
+  show (ProtoIOError filepath tag) =
+    let locS = "(" ++ Turtle.encodeString filepath ++ "): "
+        tagS = Show.shows tag "."
+     in "protobuf IO error" ++ locS ++ tagS
+  {-# INLINE show #-}
```
I generally expect `show` to give me the standard derived string output, unless it is trying to support some kind of abstraction barrier.

For custom `Exception` types I think it's better to derive `Show` and implement this pretty printing in `displayException`.

Friede80 reviewed (commented):
Still reviewing, just dropping the comments I have so far.

Friede80 reviewed (commented):
This is looking really good! I know it's still a draft, but I did run into an issue when trying to run the tests.

When I try to build there are a few modules missing from the `.cabal` file:
`Network.GRPC.MQTT.Compress`
`Network.GRPC.MQTT.Proto.TH`
`Test.Network.GRPC.MQTT.Compress`
`Test.Network.GRPC.MQTT.Compress.Gen`

Once I added them and try to run the tests I get a failure:
```
Test.Service
    Test.Service.Normal:                   OK (1.54s)
    Test.Service.ClientStream
      Test.Service.ClientStream.Unbatched: OK (0.51s)
      Batched:                             OK (1.52s)
    Test.Service.ServerStream
      Test.Service.ServerStream.Unbatched: FAIL
        Exception: GRPCIOCallError CallErrorInvalidMessage
        CallStack (from HasCallStack):
          error, called at src/Relude/Debug.hs:218:11 in relude-0.7.0.0-ygy7BwifccCX77pliNrmb:Relude.Debug
          error, called at test/Test/Service.hs:412:19 in main:Test.Service
        Use -p '/Test.Service.ServerStream.Unbatched/' to rerun this test only.
```

Friede80 commented on src/Network/GRPC/MQTT/Compress.hs:
```diff
@@ -84,47 +77,21 @@ toRemoteError err =
+-- | Decompress a 'ByteString' (compressed via 'compress') as a single
+-- zstandard frame.
+--
+-- * 'StreamError' is emitted if the 'ByteString' was compressed in streaming
+--   mode.
+--
+-- * 'GenericError' is emitted if an error was thrown by the internal zstandard
+--   FFI call.
```
Is this still true with your changes to `ZstdError`?

riz0id replied:
Thanks for catching this, I've updated the documentation for it.

Friede80 commented on src/Network/GRPC/MQTT/Message/Packet.hs:
```diff
@@ -174,22 +198,48 @@ wireParsePacket = Packet <$> wireParsePayloadField <*> wireParseMetadataField
 -- @since 0.1.0.0
-packetReader :: TChan LByteString -> ExceptT ParseError IO LByteString
-packetReader = runPacketReader loop
+makePacketReader ::
```
I'm a little confused about the `make` nomenclature. Especially given that we have a `PackerReader` data type, intuitively to me, I would expect a function called `makePacketReader` to return a `PacketReader`.

Is there some reason you chose this over something like `readPacket`?

riz0id replied:
Here `TChan LByteString -> WireDecodeOptions -> ExceptT WireDecodeError IO ByteString` is the `PacketReader` even though it is not named. The make prefix is to denote that you are constructing an thing/object that captures a computation that reads all packets into a `ByteString` and is consistent with the analogous functions for other message types like stream chunks.

riz0id commented:
> This is looking really good! I know it's still a draft, but I did run into an issue when trying to run the tests.
> 
> When I try to build there are a few modules missing from the `.cabal` file: `Network.GRPC.MQTT.Compress` `Network.GRPC.MQTT.Proto.TH` `Test.Network.GRPC.MQTT.Compress` `Test.Network.GRPC.MQTT.Compress.Gen`
> 
> Once I added them and try to run the tests I get a failure:
> 
> ```
> Test.Service
>     Test.Service.Normal:                   OK (1.54s)
>     Test.Service.ClientStream
>       Test.Service.ClientStream.Unbatched: OK (0.51s)
>       Batched:                             OK (1.52s)
>     Test.Service.ServerStream
>       Test.Service.ServerStream.Unbatched: FAIL
>         Exception: GRPCIOCallError CallErrorInvalidMessage
>         CallStack (from HasCallStack):
>           error, called at src/Relude/Debug.hs:218:11 in relude-0.7.0.0-ygy7BwifccCX77pliNrmb:Relude.Debug
>           error, called at test/Test/Service.hs:412:19 in main:Test.Service
>         Use -p '/Test.Service.ServerStream.Unbatched/' to rerun this test only.
> ```

This `CallErrorInvalidMessage` error _was_ produced for two reasons:

1. Decompressing an uncompressed payload would cause `Codec.Compression.Zstd.decompress` return a `Skip` decompression result since it sees the incoming `ByteString` lacks a compression frame header. This becomes a problem whenever the compression function chooses not to add the zstd header to the `ByteString` since that `ByteString` could not be compressed enough (or at all) such that adding a header would result in a smaller `ByteString`. The issue was fixed by choosing to return the original `ByteString` provided to `Network.GRPC.MQTT.Compress.decompress` if the result of `Zstd.decompress` is `Skip`. I don't love this solution though since `Zstd.decompress` will also return `Skip` if the `ByteString` was compressed with zstandard's streaming interface, breaking the assumption that `Skip` means the `ByteString` was not compressed in the first place. However, seeing as `grpc-mqtt` doesn't make use of any of the streaming zstd compression functions, I don't see this as an immediate issue. I've also added an additional test to ensure that decompressing an uncompressed `ByteString` is a no-op to guard against regressions in the future.

2. There were a few asymmetries in how compression and decompression got applied in the message handlers (my mistake). I changed the message handler functions by inlining `ExceptT e IO` actions that performed `TChan` reads and `Message` parsing computations. This change to the message handlers prevents them from being misused in ways that could result in compression asymmetries, double packetization, and etc... This change also makes it easy to create small, isolated roundtrip tests for each message type's send (MQTT publish) and `TChan` read functions.

ixmatus reviewed (approved):
LGTM. This was a lot of work, thank you for it!

Friede80 commented on src/Network/GRPC/MQTT/RemoteClient.hs:
```diff
@@ -195,106 +193,75 @@ handleNewSession config handle = handleWithHeartbeat
+            _ <- tryPutTMVar (hdlHeartbeat handle) ()
+            pure ()
+        Right unknown -> do
+          -- TODO: cancel hdlThread and abort
```
Do we really need to cancel the thread if we get an unknown message?

Friede80 reviewed (approved):
This is really great. Thanks @riz0id!
